### PR TITLE
[feature] Implement save and load for Mind Map CameraLookAt and CameraFollow

### LINKED ISSUE #430: Mind Map Builder: Add Camera Follows and Look At to the Builder System
### Task Scope

1. Add new field in `MindMapTreeData` to store the **position** of `Camera Follows` and `Camera LookAt`. 
2. Store positions of `Camera Follows` and `Camera LookAt` to `MindMapTreeData`
3. When loading `MindMapTree`, make sure every `MindMapNode` read and apply the value of `Camera Follows` and `Camera LookAt` to the corresponding gameobject. **Save/Load** method can be seen on `MindMapBuilder` class

Image

Structure of `MindMapNode` prefab. `MindMapNode` has `CameraFollow` and `CameraLookAt` gameobject as its children.

### References
- Intro to Scriptable Object: https://learn.unity.com/tutorial/introduction-to-scriptable-objects#
- Mind Map Builder document: https://docs.google.com/document/d/1jOUZTEIYtaI7TXRxwd3nMmKrF3J6jxNX5DeyHS1iMV4/edit?usp=sharing

### PR DESCRIPTION
**Pull request by worthlessowl**
### Main changes
Serialize CameraLookAt and CameraFollow object position for Mind Map. This PR implements the code related to saving and loading serialized value of CameraLookAt and CameraFollow.

Add ExecuteInEditMode Attribute to mindmap scripts to solve `Awake` not being called when mindmaps are manipulated in editor.

This issue affects several functions on Mindmap
- Loading saved MindMapTreeData Asset
- CameraLookAt position setter
- CameraFollow position setter

### Tasks List
[close #430]

### Attachments
Demonstrating how Load doesn't work properly because codes in the Awake function are not being called

https://user-images.githubusercontent.com/53881082/209811042-2d52b6c5-9ee4-4058-b721-9da51ece939a.mp4

Demonstrating saving and loading CameraLookAt and CameraFollow

https://user-images.githubusercontent.com/53881082/209811209-ea536dbc-2abf-4642-b628-9548332dbc00.mp4

### New Packages
```
List of new packages that you use to developed the feature(s)
```

### Resources 
```
List of url(s) used to developed the feature(s)
```

### CODE REVIEW

File: Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs [hunk 1]
```diff
@@ -1,31 +1,71 @@
...
+    public MindMapNodeType NodeType;
+    public RuntimeAnimatorController AnimController;
+    [HideInInspector]
+    public List<MindMapNode> Children = new List<MindMapNode>();
+    public bool IsDiscovered = false;
+    public MindMapNode Parent = null;
+
+    [SerializeField]
+    [Tooltip("Position relative to node")]
+    private Vector3 _cameraStartingPosition;
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs:33]: we dont use this field, remove this field together with its decorator

File: Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs [hunk 2]
```diff
@@ -1,31 +1,71 @@
...
+    [SerializeField]
+    [Tooltip("Degree around each axis")]
+    private Vector3 _cameraStartingAngle;
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs:36]: we dont use this field, remove this field together with its decorator

File: Assets/Astralization/Scripts/UI/MindMap/MindMapBuilder.cs [hunk 3]
```diff
@@ -67,7 +68,11 @@ private void SaveTree(string filename)
             mindMapTreeData.NodeDescriptions.Add(currentNode.NodeDescription);
             mindMapTreeData.NodeTypes.Add(currentNode.NodeType);
             mindMapTreeData.NodeAnimationControllers.Add(currentNode.AnimController);
-            mindMapTreeData.NodeTransforms.Add(currentNode.transform);
+            mindMapTreeData.NodePositions.Add(currentNode.transform.position);
+            mindMapTreeData.NodeRotations.Add(currentNode.transform.rotation);
+            mindMapTreeData.NodeScales.Add(currentNode.transform.localScale);
+            mindMapTreeData.NodeCameraFollowPosition.Add(currentNode.GetCameraFollow().localPosition) ;
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/UI/MindMap/MindMapBuilder.cs:74]: remove whitespace before ;

File: Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs [hunk 4]
```diff
@@ -1,31 +1,71 @@
...
+using UnityEngine;
+
+public enum MindMapNodeType
+{
+    CHAPTER,
+    CORE,
+    CLUE
+}
+
+[ExecuteInEditMode]
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs:13]: MindMapNode is actually intended for runtime, not limited on builder purpose so I think this tag will make the editor heavier in the future. Use null checker on SetGet instead. When the field is null, use transform get first.

File: Assets/Astralization/Scripts/UI/MindMap/MindMapTree.cs [hunk 5]
```diff
@@ -17,21 +19,42 @@ public interface IMindMapTree
...
+#region SerializableClass
+[Serializable]
+public class NodeModelDictionary : SerializableDictionary<MindMapNodeType, GameObject> { }
+#endregion
 
 /**
  * Class for managing and manipulating mind map tree.
  * Current intended usage: Chapter Clues implementation 
  */
+[ExecuteInEditMode]
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/UI/MindMap/MindMapTree.cs:31]: similar to the MindMapNode tag, use null checker on SetGet instead.

**Review by rafiftaris — CHANGES_REQUESTED** [review]
Note: Thank you very much! I just learned another useful tag `[ExecuteInEditMode]` for that is very suitable for developing builders.

**Review by naufalhilmi12345 — APPROVED** [review]
LGTM for implementation. for code agree with rafif.

File: Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs (hunk 1 above)
Comment (worthlessowl) [reply Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs]: Addressed on 432d630c9c1a56d51d7092dba3376b7fbdda5392

File: Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs (hunk 2 above)
Comment (worthlessowl) [reply Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs]: 432d630c9c1a56d51d7092dba3376b7fbdda5392

File: Assets/Astralization/Scripts/UI/MindMap/MindMapBuilder.cs (hunk 3 above)
Comment (worthlessowl) [reply Assets/Astralization/Scripts/UI/MindMap/MindMapBuilder.cs]: addressed on 6322f656a4330b1ddb922f3167f00579ace10be0

File: Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs (hunk 4 above)
Comment (worthlessowl) [reply Assets/Astralization/Scripts/UI/MindMap/MindMapNode.cs]: Addressed on 432d630c9c1a56d51d7092dba3376b7fbdda5392

File: Assets/Astralization/Scripts/UI/MindMap/MindMapTree.cs (hunk 5 above)
Comment (worthlessowl) [reply Assets/Astralization/Scripts/UI/MindMap/MindMapTree.cs]: Addressed on 432d630c9c1a56d51d7092dba3376b7fbdda5392

File: Assets/Astralization/UI/ChapterClues/MindMapNode.prefab [hunk 6]
```diff
@@ -1,5 +1,65 @@
...
+  m_IsActive: 1
+--- !u!4 &4538325078790542776
+Transform:
+  m_ObjectHideFlags: 0
+  m_CorrespondingSourceObject: {fileID: 0}
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  m_GameObject: {fileID: 4445169330724473857}
+  m_LocalRotation: {x: 0, y: 0, z: 0, w: 1}
+  m_LocalPosition: {x: 0, y: -5, z: 0}
```
Comment (rafiftaris) [inline Assets/Astralization/UI/ChapterClues/MindMapNode.prefab:57]: change the y value to 0. change this transform of `CameraFollow` gameobject by opening the `MindMapNode` prefab

**Review by rafiftaris — CHANGES_REQUESTED** [review]
the camera fields of the builder works perfectly but dont forget to update [this docs](https://docs.google.com/document/d/1jOUZTEIYtaI7TXRxwd3nMmKrF3J6jxNX5DeyHS1iMV4/edit?usp=sharing). I'll approve immediately after the docs and this requested change are resolved

File: Assets/Astralization/Scripts/UI/MindMap/MindMapTree.cs [hunk 7]
```diff
@@ -185,6 +210,12 @@ public void LoadTree()
             nodeModel.transform.parent = newNode.transform;
             nodeModel.gameObject.layer = LayerMask.NameToLayer(LayerName);
 
+            // Type-related assignments
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/UI/MindMap/MindMapTree.cs:213]: duplicated after merging

**Review by rafiftaris — CHANGES_REQUESTED** [review]

**Review by rafiftaris — APPROVED** [review]
LGTM!

**Pull request merged**